### DISCUSSION TITLE
Compute and track the ratio of the current timestep to the previous

### DISCUSSION THREAD

**Discussion by TLWise** [idaholab/moose#18550] [Q&A General] [answered]
For a time dependent solution of u, I would like to store the previous value of u and compare it to the current value of u. Another method would be to store the current value then compare it to the a subsequent u when the value of the subsequent u is greater than or less than a magnitude of the previously stored u. Is there a simple way to do this in MOOSE without having to create a C++ code using something like a "do loop" ?

**Comment by GiudGiud** [comment]
Would any of these work for you:
https://mooseframework.inl.gov/source/postprocessors/ChangeOverTimePostprocessor.html
https://mooseframework.inl.gov/source/postprocessors/ChangeOverTimestepPostprocessor.html

**Reply by TLWise** [reply]
Looks like it should do the trick. I will test it and provide an update on the result.

Thank you.

**Reply by TLWise** [reply]
It would appear that using the ChangeOverTime or Timestep would work in conjunction with SolutionFunction. The problem is that when I use solution function as defined:

  [./solution_fcn]
    type = SolutionFunction
    from_variable = u
    solution = solution_uo
  [../]


I get the following error: (Functions/solution_fcn/solution):  The requested UserObject with the name "solution_uo" was not found.

**Reply by GiudGiud** [reply]
do you have a solution_uo somewhere? If not you need to add it
But I don't understand what the solution function is here for

**Reply by TLWise** [reply]
I copied this function from an example problem for SolutionFunction. The current site for it is under construction and there was not any real good instructions beyond the examples provided. There was also no other reference to the solution_uo. This function (SolutionFunction)seems to be the only one that I could find that will operate on a kernel variable.

**Reply by GiudGiud** [reply]
SolutionFunction is used to have the result of a previous solve, eg loaded using a solution user object, in a function. 
It's not really for looking at older values within the same solve. 

The two post processors I linked do that. What are you trying to get from this comparison of current and previous value? Examine convergence ?

**Reply by TLWise** [reply]
Yes. I am looking to have a means to measure convergence/divergence of the current timestep solution to a reference timestep. The output would be in the form of the resulting ratio.

**Reply by GiudGiud** [reply]
Do you already know the solution at the reference timestep? Or are you computing it during the simulation (is it like the last one?)

**Reply by TLWise** [reply]
The reference value is stored in solution during the solve. The first time step solution of u is the initial reference solution. The current solution is compared to the reference u and then becomes the new reference u etc.

**Reply by GiudGiud** [reply]
Does it always become the new reference? If so then ChangeOverTimestep compares to the previous one. 
If not, then I'd write a custom aux kernel based on SelfAux to save the reference value, then compare with a DifferencePostprocessor or a ParsedPostprocessor

**Reply by TLWise** [reply]
Yes it always become the new reference. Can ChangeOverTimestep be apples directly to a kernel variable u. It seems from the example for use of ChangeOverTimestep that it works only with a defined function. What would be the syntax required to have ChangeOverTimestep operate on the variable u?

**Reply by GiudGiud** [reply]
Ah you are right these are totally not the right postprocessors to use to compare variables directly.

**Comment by GiudGiud** [comment, marked as answer]
So I'd recommend:
1) a SelfAux to save the variable value at TIMESTEP_BEGIN, to have the old value:
https://mooseframework.inl.gov/source/auxkernels/SelfAux.html

2) Then a ParsedAux to compute the difference, relative difference or whatever you need, and store it in another aux variable, at TIMESTEP_END for example. 

3) Then a ElementL1Error comparing the new aux variable to a 0 reference function. 
https://mooseframework.inl.gov/source/postprocessors/ElementL1Error.html

If you want an L2 error or something that already exists, you can skip step 2-3 and just use that. 

Or you could write your own postprocessor to do this. In which case, you can get old values of variables directly in the postprocessor using coupledValueOld
